C_EXTENSION.md

```diff
@@ -0,0 +1,135 @@
+
+C extension is built under `ion-python/amazon/ion` and named according to the following format (may be slightly different depending on your platform) `ionc.cpython-$py_version-$platform.$suffix` (e.g., ionc.cpython-39-darwin.so)
+
+#### Getting Started with C Extension:
+```
+>>> import amazon.ion.simpleion as ion
+>>> obj = ion.loads('{abc: 123}')
+>>> obj['abc']
+123
+>>> ion.dumps(obj, binary=True)
+b'\xe0\x01\x00\xea\xe9\x81\x83\xd6\x87\xb4\x83abc\xd3\x8a!{'
+```
+
+
+## Development
+
+Architecture of Ion Python C extension:
+```
+                                   ioncmodule.c
+                                        |
+                                        | 
+                                        ↓ 
+Ion C -------> Ion C binaries -----> setup.py ------> C extension -------------------> Ion Python simpleion module
+      compile                                  setup               import ionc module   
+```
+After setup, C extension will be built and imported to simpleion module. If there are changes in `ioncmodule.c`, build the latest C extension by running `python setup.py build_ext --inplace`.
+
+
+## Technical Details
+
+### 1. Common Binary Encoding Differences between C Extension and Original Ion Python
+Note that both binary encodings are **equivalent**; one encoding is not more "correct" than the other.<br/>
+
+#### 1.1 Different ways to represent a struct's length. Refer to [Amazon Ion Binary Encoding](https://amzn.github.io/ion-docs/docs/binary.html#13-struct) for details.<br/>
+For Ion struct `{a:2}`:
+```text
+Text       IVM               ion_symbol_table::{         symbols:[”a”]}  {         “a”: 2     }   
+Ion C      \xe0\x01\x00\xea  \xe7\x81\x83      \xd4      \x87\xb2\x81a   \xd3      \x8a 21\x02
+Ion Python \xe0\x01\x00\xea  \xe8\x81\x83      \xde\x84  \x87\xb2\x81a   \xde\x83  \x8a 21\x02
+```
+
+#### 1.2 Different order of symbols within a symbol table.<br/>
+For symbol `abc` with two annotations `annot1` and `annot2`, `annot1::annot2::abc`:
+```text
+Ion C text        ion_symbol_table::{         symbols:[    "abc", "annot1", "annot2"]}          annot1($11)::annot2($12)::abc($10)
+Ion C binary      \xee\x99\x81\x83  \xde\x95  \x87\xbe\x92 \x83abc\x86annot1\x86annot2 \xe5\x82 \x8b         \x8c         \x71\x0a
+Ion Python binary ion_symbol_table::{         symbols:[    "annot1", "annot2", "abc",]}         annot1($10)::annot2($11)::abc($12)
+ion Python        \xee\x99\x81\x83  \xde\x95  \x87\xbe\x92 \x86annot1\x86annot2\x83abc \xe5\x82 \x8a         \x8b         \x71\x0c
+```
+
+### 2. Known Issues
+1. Type translation errors causing python to crash. Refer to [amzn/ion-python#151](https://github.com/amzn/ion-python/issues/151) for details.
```

Comment:
Is this still an issue?

Reply:
Thanks, I barely see this issue now. I also closed #151 earlier. I'll update the document.